### ISSUE TITLE
Android `setOrientationBis` chooses orientation with `w > h` instead of sensor

### ISSUE THREAD

**Issue by AntTheAlchemist** [libsdl-org/SDL#10377]
In the case of `SDL_SetHint(SDL_HINT_ORIENTATIONS, "LandscapeLeft LandscapeRight Portrait PortraitUpsideDown")`, `setOrientationBis` will decide depending on w > h. This is a bug. It should be using `SCREEN_ORIENTATION_FULL_SENSOR` instead, to let the sensor decide.

**Comment by slime73** [comment]
Looking at the code, it seems like it will do what you want if the window is resizable, right? Here: https://github.com/libsdl-org/SDL/blob/a880410cb2cc03d9e3b88c358f79f0eb0e7e7c4e/android-project/app/src/main/java/org/libsdl/app/SDLActivity.java#L1142-L1146

**Comment by AntTheAlchemist** [comment]
> Looking at the code, it seems like it will do what you want if the window is resizable, right?

You're right - I didn't notice I was missing that flag in my window creation 🤦🏼‍♂️  Thanks!

**Comment by AntTheAlchemist** [comment]
I did find another bug though. If choosing between a flipped, or not, version of landscape / portrait, the user setting isn't respected, so it will always use the sensor, regardless of user's auto rotate setting. So, if it's a full 360 rotation, that works fine, and respects the user's auto rotate. But landscapeLeft & landscapeRight will always be sensor, ignoring user's auto rotate. Does that make sense?

**Comment by slouken** [comment]
Patches welcome!

**Comment by AntTheAlchemist** [comment]
[07636ac](https://github.com/libsdl-org/SDL/pull/10394/commits/07636acd071e35102b35893cf03416ef485a830f) fixes the user's auto-rotate issue, but I found yet another problem in the logic:

`SDL_SetHint(SDL_HINT_ORIENTATIONS, "LandscapeLeft LandscapeRight Portrait");` will allow PortraitUpsideDown, because there's no distinction between one portrait and both landscape (and visa versa... one landscape and both portraits) when deciding to allow all orientations (hope that makes sense). I don't see a way to actually fix this without looking to see if we're attempting to get landscape or portrait. `setOrientationBis()` is flawed; a different approach is needed.

**Comment by slouken** [comment]
> OK. I found out one weird thing. When on Android, somehow in landscape mode, the width is large while the height is small. So basically, it's providing a viewport of a landscape window, even though the window being rendered is actually portrait (I see it being cut off in this mode, as the horizontal vs vertical sizes are inversed in the window when created without any sizes (for autodetection using the display it's connected to))?

Ah, can you update to the latest main code and see if that's fixed?

**Comment by superfury** [comment]
Didn't update yet, but quickly looked for related commits since 3.1.6 prerelease.

I did see some function change to a different function call for rendering surfaces, but the main issue is weirder than that.

Basically, from what I can see stepping through the code, the display ID (1) used by the window creation (and all steps including the renderer and texture) without x/y resolution specified reports the horizontal and vertical axes flipped in landscape mode. That basically puts any textures created based on the window out-of-range.

I think it's because for example a 1K x 2K portrait screen reports 1K x 2K in portrait mode (based on orientation events) and 2K x 1K in landscape mode.
The issue is that everything is always rendered in portrait mode, even in 'landscape mode'.
So it creates a 2K x 1K texture and surface and writes those pixels to a 1K x 2K framebuffer in portrait mode. I see a 4K screen getting (texture rendered at to +90 degrees angle rendering by my app due to landscape mode) rendered with what looks like half the screen vertically centered in landscape mode and the right half of the rendered display being off-screen.
So it's like it's mapping the right half out-of-frame (clipping it or corrrupting memory?) and only the top 1/2 of the surface rendered, with roughly 25% black bar above and 25% black bar below (at a ridiculously low resolution)?
The renderer is set to use letterboxing and uses SDL_GetDisplayUsableBounds to obtain a resolution to try and set using SDL_setWindowSize (the source of the incorrect axes in landscape mode).

**Comment by superfury** [comment]
I'll check it out when I have time.

Also, is SDL_GetDisplayUsableBounds also fixed to report correctly?
I seem to notice stepping through the code on my Android phone that the SDL_setWindowSize function (based on the above function call on 3.1.6) doesn't have a handler installed on _this->SetWindowSize on Android, reaching SDL_Unsupported(), located on line 65 right now ? Is that an issue?
I think I remember the same on the SDL_SetWindowPosition call too.

**Comment by superfury** [comment]
Just got the current git commit (3.1.7), adjusted my code to no longer flip or orientate using textures (just back to the old SDL_RenderTexture, using a on-render SDL_CreateTextureFromSurface (and calling SDL_DestroyTexture after rendering, before calling SDL_RenderPresent).

The screen now displays correctly on all orientations (using the current SDL3 commit).

**Comment by slouken** [comment]
> The screen now displays correctly on all orientations (using the current SDL3 commit).

Great, thanks for the confirmation!

**Comment by AntTheAlchemist** [comment]
@superfury I'm still seeing issue #11324 on opengles2 with current source. The first present after rotation will draw the frame using distorted coordinates. Landscape to portrait seems to be drawn half-way up the screen. Portrait to landscape is all over the place - cannot tell you what's trying to be drawn. Then everything is drawn okay after the first bad present. You'll miss it visually if doing a present every iteration. I don't know if this will mess up your observations.

**Comment by superfury** [comment]
@AntTheAlchemist I indeed see that happening on current commit with Windows SDL3 builds (both in windowed mode (with letterbox) and fullscreen mode with letterbox).
Android otoh looks like it's running without issues, although perhaps with a slowdown (but I'm not sure if that's due to SDL or my own app).

**Comment by AntTheAlchemist** [comment]
@superfury okay that's strange because I'm only seeing it on Android. Just tried it on a Windows 10 tablet and rotation works okay first frame. I can only say for visual drawing coordinates, not mouse coordinates - not set up a test for that. What's going on here? Android's Vulkan renderer doesn't have this problem, only opengles2.

**Comment by superfury** [comment]
Mouse coordinates are fine from what I can see (using SDL_ConvertEventToRenderCoordinates).
The main issue is that the surface when rendered using the SDL1-2-3 conversion process (using below code actually, with SDL3 being defined) is invalidly rendering the first frame after the resolution change.
```
//SDL2/3!
// Select the color for drawing. It is set to black here.
SDL_SetRenderDrawColor(sdlRenderer, 0x00, 0x00, 0x00, SDL_ALPHA_OPAQUE);
// Clear the entire screen to our selected color.
SDL_RenderClear(sdlRenderer);
// Copy over our display!
if (sdlTexture) //Gotten a valid texture?
{
				SDL_UpdateTexture(sdlTexture, NULL, *getlayerpixels(surface),(get_pixelrow_pitch(surface) << 2)); //Update the texture!

				#ifdef SDL3
				SDL_RenderTexture(sdlRenderer, sdlTexture, NULL, NULL); //Default behaviour!
				#else
				SDL_RenderCopy(sdlRenderer, sdlTexture, NULL, NULL);
				#endif
}

// Up until now everything was drawn behind the scenes.
// This will show the new contents of the window.
SDL_RenderPresent(sdlRenderer);
```

getlayerpixels(sdllayer) gives a pointer as &sdllayer->pixels, so effectively boils down to sdllayer->pixels passed to SDL_RenderTexture
get_pixelrow_pitch(sdllayer) returns sdllayer->pitch divided by 4 if 4 or higher, layer->w otherwise (if it's less than 4, just a safety value, usually doesn't happen as it's using doubleword (32-bit) integer pixels)).

**Comment by ObaraEmmanuel** [comment]
@slouken my android app works in `sensorLandscape` but somewhere along the way to version 3.2.14 it stopped working. I tracked down the issue to `setOrientationBis`. The special condition that allowed `sensorLandscape` to work is that the `SDL_HINT_ORIENTATIONS` hint is not specified and that width > height. When I checked using a debugger, `setOrientationBis` is called the first time around using the right screen dimensions and `sensorLandscape` is set. However, almost immediately, it is called again with width = 1 and height = 1 which sets the orientation to `sensorPortrait` and screws everything up.

My window creation is as follows:

```c
ctx->window = SDL_CreateWindow(
        "NES Emulator",
        0,
        0,
        SDL_WINDOW_FULLSCREEN
        | SDL_WINDOW_HIGH_PIXEL_DENSITY
    );
```

My android manifest is set to use `sensorLandscape`.

I figured this issue is very closely related to my problem.

**Comment by slouken** [comment]
I've gone ahead and ignored setOrientationBis() with width/height of 1. I think this takes care of all the issues in this bug. Please open a new bug with details if you're seeing any further issues.
